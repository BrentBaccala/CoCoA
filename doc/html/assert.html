<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML>
<HEAD>
<META NAME="generator" CONTENT="http://txt2tags.org">
<LINK REL="stylesheet" TYPE="text/css" HREF="cocoalib-doc.css">
<TITLE>      assert</TITLE>
</HEAD><BODY BGCOLOR="white" TEXT="black">
<CENTER>
<H1>      assert</H1>
<FONT SIZE="4"><I>      &copy;  2007 John Abbott</I></FONT><BR>
<FONT SIZE="4">      GNU Free Documentation License, Version 1.2</FONT>
</CENTER>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>

    <UL>
    <LI><A HREF="#examples">Examples</A>
    <LI><A HREF="#toc2">User documentation for files assert.H and assert.C</A>
      <UL>
      <LI><A HREF="#toc3">Debugging</A>
      </UL>
    <LI><A HREF="#toc4">Maintainer documentation for files assert.H and assert.C</A>
    <LI><A HREF="#toc5">Bugs, Shortcomings, and other ideas</A>
    </UL>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>
<P>
      <center><a href="index.html">CoCoALib Documentation Index</a></center>
</P>

<A NAME="examples"></A>
<H2>Examples</H2>

<UL>
<LI><A HREF="../../examples/index.html#ex-PolyIterator2.C">ex-PolyIterator2.C</A>
</UL>

<A NAME="toc2"></A>
<H2>User documentation for files assert.H and assert.C</H2>

<P>
The only part of <CODE>assert.H</CODE> which a normal CoCoALib user might find useful
is the <CODE>CoCoA_ASSERT</CODE> macro: it is intended to be a debugging aid.
</P>
<P>
The <CODE>CoCoA_ASSERT</CODE> macro does absolutely nothing (not even evaluating its
argument) unless the compilation flag <CODE>CoCoA_DEBUG</CODE> is set.  If that flag
is set then the macro evaluates its argument to a boolean result which
is then tested: if the result is true nothing further happens, if the
result is false then the function <CODE>CoCoA::AssertionFailed</CODE> is called with
some arguments indicating which <CODE>CoCoA_ASSERT</CODE> macro call obtained the false
value.  The <CODE>AssertionFailed</CODE> function prints out an informative message on
<CODE>std::cerr</CODE> and then throws a <CODE>CoCoA::ERR::AssertFail</CODE> exception.
</P>
<P>
The file <CODE>assert.H</CODE> contains the following:
-  definition of the <CODE>CoCoA_ASSERT</CODE> macro to aid debugging, and the
     related function <CODE>AssertionFailed</CODE>
</P>

<A NAME="toc3"></A>
<H3>Debugging</H3>

<P>
During debugging, a debugger can be used to intercept calls to the
function <CODE>CoCoA::AssertionFailed</CODE> which will stop the program just before
throwing the <CODE>CoCoA::ERR::AssertFail</CODE> exception.  This should enable one
to find more easily the cause of the problem.
</P>
<P>
For example, in <B>gdb</B> type
</P>

<PRE>
  break CoCoA::AssertionFailed
</PRE>

<P>
and then go <CODE>up</CODE> (perhaps repeatedly) to the offending line.
</P>

<A NAME="toc4"></A>
<H2>Maintainer documentation for files assert.H and assert.C</H2>

<P>
The macro name <CODE>CoCoA_ASSERT</CODE> is rather cumbersome, but must contain the
prefix <CODE>CoCoA_</CODE> since macro names cannot be placed in C++ namespaces.
The two definitions of the macro (debugging and non-debugging cases)
both look rather clumsy, but are done that way so that the macro expands
into an expression which is syntactically a simple command.  The
definition for the non-debugging case I took from <CODE>/usr/include/assert.h</CODE>;
I do not recall where I got the definition for the debugging case, but
the definition in <CODE>/usr/include/assert.h</CODE> looked to be gcc specific.
</P>
<P>
The purpose of the procedure <CODE>AssertionFailed</CODE> is explained above in the
user documentation (to facilitate interception of failed assertions).  The
procedure never returns; instead it throws a CoCoALib exception with code
<CODE>ERR::AssertFail</CODE>.  Before throwing the exception it prints out a message
on <CODE>std::cerr</CODE> summarising what the assertion was, and where it was.
Note the non-standard way of throwing the CoCoA exception: this allows the
<CODE>ErrorInfo</CODE> object to refer to the file and line where <CODE>CoCoA_ASSERT</CODE> was
called (rather then to the line in <CODE>assert.C</CODE> where <CODE>CoCoA_ERROR</CODE> is called).
The entire printed message is assembled into an <CODE>ostringstream</CODE> before being
printed to ensure exception safety: either the whole message is printed or
none of it is, since the printing step is an atomic operation.
</P>

<A NAME="toc5"></A>
<H2>Bugs, Shortcomings, and other ideas</H2>

<P>
Is the exception safe implementation of <CODE>AssertionFailed</CODE> excessive?
</P>
<P>
You have to use explicitly <CODE>#ifdef CoCoA_DEBUG</CODE> if you want to have a
loop or any other non-trivial piece of code executed only when debugging
it turned on.
</P>
<P>
The following (simplified but real) code excerpt is mildly problematic:
</P>

<PRE>
  {
    bool OK = ....;
    CoCoA_ASSERT(OK);
  }
</PRE>

<P>
When compiled without debugging (<I>i.e.</I> <CODE>CoCoA_DEBUG</CODE> is zero) the compiler
(gcc-3) complains that the variable <CODE>OK</CODE> is unusued.  It does not appear
to be possible to make the macro "depend on its argument" in the
non-debugging case without incurring the run-time cost of evaluating the
argument (if the argument is just a variable the cost is negligible, but
if it is a more complex expression then the cost could be considerable).
The solution adopted was to modify the calling code like this:
</P>

<PRE>
  {
    bool OK;
    OK = ....;
    CoCoA_ASSERT(OK);
  }
</PRE>

<P>
Note that the apparently simpler code below <B>will not work</B> if the elided
code (<I>i.e.</I> the <CODE>....</CODE>) has a side effect since the elided code will not
be called at all in the non-debugging case:
</P>

<PRE>
  {
    CoCoA_ASSERT(....);
  }
</PRE>

<P>
<B>POSSIBLE SOLUTION</B>: maybe <CODE>CoCoA_ASSERT</CODE> could compute <CODE>sizeof(...)</CODE> in the
non-debugging case -- this should avoid evaluation of the argument, and
will compile away to nothing.
</P>

<!-- html code generated by txt2tags 2.6 (http://txt2tags.org) -->
<!-- cmdline: txt2tags -t html -o html/assert.html -i txt/assert.txt -->
</BODY></HTML>
