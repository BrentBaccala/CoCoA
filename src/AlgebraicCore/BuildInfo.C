//   Copyright (c)  2005,2007,2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/config.H"

#include <limits>
#include <iostream>
using std::ostream;
using std::endl;
#include <string>
using std::string;

namespace CoCoA
{

  namespace BuildInfo
  {
    // Build info is actually passed in via three preprocessor variables:
    // COCOA_VERSION, COCOA_CXX, COCOA_CXXFLAGS -- all three are quoted strings.

    const std::string& version()
    {
      static const string info(COCOA_VERSION);
      return info;
    }

    const std::string& compiler()
    {
      static const string info(COCOA_CXX);
      return info;
    }

    const std::string& CompilationFlags()
    {
      static const string info(COCOA_CXXFLAGS);
      return info;
    }

    const std::string& CompilationDefines()
    {
      static const string info(COCOA_CXXFLAGS_DEFINES);
      return info;
    }


    void PrintAll(std::ostream& out)
    {
    // Below we use string literal juxtaposition, so that the build
    // info can readily be extracted from libcocoa.a by the command
    // strings libcocoa.a | egrep "CoCoA::BuildInfo"
      out << endl
          << "CoCoA::BuildInfo Summary of build information for CoCoALib:" << endl
          << "CoCoA::BuildInfo CoCoALib Version: " COCOA_VERSION << endl
          << "CoCoA::BuildInfo Compiler: " COCOA_CXX << endl
          << "CoCoA::BuildInfo Compilation Flags: " COCOA_CXXFLAGS << endl
          << "CoCoA::BuildInfo Compilation Flags Defines: " COCOA_CXXFLAGS_DEFINES << endl
          << "CoCoA::BuildInfo NumBits int = " << std::numeric_limits<unsigned int>::digits << endl
          << "CoCoA::BuildInfo NumBits long = " << std::numeric_limits<unsigned long>::digits << endl
          << "CoCoA::BuildInfo NumBits SmallExponent_t = " << std::numeric_limits<SmallExponent_t>::digits << endl
          << endl;
    }

  } // end of namespace CoCoA::BuildInfo

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/BuildInfo.C,v 1.8 2012/10/24 12:10:45 abbott Exp $
// $Log: BuildInfo.C,v $
// Revision 1.8  2012/10/24 12:10:45  abbott
// Revised PrintAll: now prints also NumBits of SmallExponent_t
//
// Revision 1.7  2012/10/02 11:45:09  bigatti
// -- added CompilationDefines
//
// Revision 1.6  2012/10/02 10:36:12  abbott
// Revised interface to BuildInfo information strings.
// Several consequential changes.
//
// Revision 1.5  2012/06/20 12:26:26  bigatti
// -- now exporting also compiler and compilation flags
//
// Revision 1.4  2011/03/11 17:38:21  bigatti
// -- added missing include
//
// Revision 1.3  2011/03/11 17:18:55  bigatti
// -- added info on int and long size
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.4  2007/02/10 18:44:03  cocoa
// Added "const" twice to each test and example.
// Eliminated dependency on io.H in several files.
// Improved BuildInfo, and added an example about how to use it.
// Some other minor cleaning.
//
// Revision 1.3  2007/02/08 22:34:22  cocoa
// Changed BuildInfo: only BuildInfo version is publicly visible.
// Only BuildInfo needs complicated compile-time flags, so several changes
// to Makefiles etc.  Added a new example: ex-BuildInfo.C
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.2  2005/10/12 15:52:09  cocoa
// Completed test-RingFp1 and corrected/cleaned the SmallFp*
// and RingFp* files.
//
// Some minor tidying elsewhere.
//
// Revision 1.1  2005/10/06 16:36:42  cocoa
// Added the capability find out build information at run-time.
// The Makefiles should be a little tidier too.
//
