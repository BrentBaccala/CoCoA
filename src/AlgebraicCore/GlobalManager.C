//   Copyright (c)  2007,2010,2011  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/GlobalManager.H"
#include "CoCoA/FractionField.H"
#include "CoCoA/MemPool.H"
#include "CoCoA/DenseUPolyRing.H" // for Hilbert-Poincare' series
#include "CoCoA/utils.H"

#include "CoCoA/error.H"
#include "TmpHilbertDir/TmpPoincareCPP.H" // for Hilbert-Poincare' series

#include "gmp.h"

#include <algorithm>
using std::min;
using std::max;
#include <cstdlib>
using std::malloc;
using std::realloc;
using std::free;
#include <iostream>
// using std::cerr & std::endl for serious warning in GlobalManager dtor
#include <cstring>
using std::memcpy;


// These 3 fns must have C linkage to work with GMP's mem mgr setter.
extern "C"
{
  void* CoCoA_GMP_alloc(size_t sz);
  void* CoCoA_GMP_realloc(void* ptr, size_t oldsz, size_t newsz);
  void CoCoA_GMP_free(void* ptr, size_t sz);
}

void* CoCoA_GMP_alloc(size_t sz)
{
  CoCoA_ASSERT(CoCoA::GlobalGMPSliceSize() != 0);
  if (sz <= CoCoA::GlobalGMPSliceSize())
    return CoCoA::GlobalGMPPoolPtr()->alloc();
  return malloc(sz);
}

void* CoCoA_GMP_realloc(void* ptr, size_t oldsz, size_t newsz)
{
  CoCoA_ASSERT(CoCoA::GlobalGMPSliceSize() != 0);
  if (oldsz <= CoCoA::GlobalGMPSliceSize() &&
      newsz <= CoCoA::GlobalGMPSliceSize())
    return ptr;

  if (oldsz > CoCoA::GlobalGMPSliceSize() &&
      newsz > CoCoA::GlobalGMPSliceSize())
    return realloc(ptr, newsz);

  const size_t n = min(oldsz, newsz);
  void* dest = CoCoA_GMP_alloc(newsz);
  memcpy(dest, ptr, n);
  CoCoA_GMP_free(ptr, oldsz);
  return dest;
}

void CoCoA_GMP_free(void* ptr, size_t sz)
{
  CoCoA_ASSERT(CoCoA::GlobalGMPSliceSize() != 0);
  if (sz <= CoCoA::GlobalGMPSliceSize())
    CoCoA::GlobalGMPPoolPtr()->free(ptr);
  else
    free(ptr);
}



namespace CoCoA
{

  // Pseudo-ctors for RingZZ and RingQ.
  ring MakeUniqueInstanceOfRingZZ(); // Defined in RingZZ.C.
  FractionField MakeUniqueInstanceOfRingQQ(const ring&); // Defined in RingQ.C.

  // Checking fns to be called immediately before calling dtors for RingQ and RingZZ
  bool RingZZStillInUse(const ring& ZZ);  // Defined in RingZZ.C
  bool RingQQStillInUse(const FractionField& Q);  // Defined in RingQ.C


  // The static members of GlobalManager -- effectively global variables.
  const GlobalManager* GlobalManager::ourGlobalDataPtr = 0;
  std::size_t GlobalManager::GMPSliceSize = 0; // size in bytes of slices in the MemPool (compile-time constant)
  MemPool* GlobalManager::GMPPoolPtr = 0;
  std::size_t GlobalManager::ourHPMaxPower = 100;  // for Hilbert-Poincare' series

  const GlobalManager* GlobalManager::ptr(const char* FnName)
  {
    if (GlobalManager::ourGlobalDataPtr == 0)
      CoCoA_ERROR(ERR::NoGlobalMgr, FnName);
    return GlobalManager::ourGlobalDataPtr;
  }


  GlobalManager::GMPMemMgr::GMPMemMgr(GlobalSettings::AllocatorSetting choice, std::size_t SliceSize)
  {
    if (choice == GlobalSettings::SystemAllocator) return;

    myPoolPtr.reset(new MemPool(SliceSize, "Global GMP MemPool")); // must do this first to be exception safe
    GlobalManager::GMPPoolPtr = myPoolPtr.get();
    GlobalManager::GMPSliceSize = GlobalManager::GMPPoolPtr->mySliceSize();
    mp_get_memory_functions(&myPrevAlloc, &myPrevRealloc, &myPrevFree);
    mp_set_memory_functions(&CoCoA_GMP_alloc, &CoCoA_GMP_realloc, &CoCoA_GMP_free);
  }


  GlobalManager::GMPMemMgr::~GMPMemMgr()
  {
    if (myPoolPtr.get() == 0) return;

    mp_set_memory_functions(myPrevAlloc, myPrevRealloc, myPrevFree);
    GlobalManager::GMPSliceSize = 0;
    GlobalManager::GMPPoolPtr = 0;
  }


  // ----------------------------------------------------------------------

  const std::size_t GlobalSettings::ourDefaultSliceSize = 2*sizeof(long);
  const GlobalSettings::ResidueSetting GlobalSettings::ourDefaultResidueSetting = GlobalSettings::SymmResidues;
  const GlobalSettings::AllocatorSetting GlobalSettings::ourDefaultAllocatorSetting = GlobalSettings::SystemAllocator;


  GlobalSettings::GlobalSettings():
      myResidueSettingHasBeenSet(false),
      myAllocatorSettingHasBeenSet(false),
      mySliceSizeHasBeenSet(false),
      myResidueSetting(ourDefaultResidueSetting),
      myAllocatorSetting(ourDefaultAllocatorSetting),
      mySliceSize(ourDefaultSliceSize)
  {}

  GlobalSettings& GlobalSettings::mySetResidueSetting(ResidueSetting r)
  {
    CoCoA_ASSERT(!myResidueSettingHasBeenSet);
    myResidueSettingHasBeenSet = true;
    myResidueSetting = r;
    return *this;
  }

  GlobalSettings& GlobalSettings::mySetAllocatorSetting(AllocatorSetting a)
  {
    CoCoA_ASSERT(!myAllocatorSettingHasBeenSet);
    myAllocatorSettingHasBeenSet = true;
    myAllocatorSetting = a;
    return *this;
  }

  GlobalSettings& GlobalSettings::mySetSliceSize(std::size_t SliceSize)
  {
    CoCoA_ASSERT(!mySliceSizeHasBeenSet);
    mySliceSizeHasBeenSet = true;
    mySliceSize = SliceSize;
    return *this;
  }



  GlobalSettings GlobalSettings::operator()(std::size_t SliceSize) const
  {
    CoCoA_ASSERT(!mySliceSizeHasBeenSet && myAllocatorSetting != SystemAllocator);
    GlobalSettings ans(*this);
    return ans.mySetSliceSize(SliceSize);
  }


  GlobalSettings operator+(const GlobalSettings& arg1, const GlobalSettings& arg2)
  {
    GlobalSettings ans;
    if (arg1.myResidueSettingHasBeenSet && arg2.myResidueSettingHasBeenSet)
      CoCoA_ERROR(ERR::BadGlobalSettings, "residue setting");
    if (arg1.myResidueSettingHasBeenSet) ans.mySetResidueSetting(arg1.myResidueSetting);
    if (arg2.myResidueSettingHasBeenSet) ans.mySetResidueSetting(arg2.myResidueSetting);

    if (arg1.myAllocatorSettingHasBeenSet && arg2.myAllocatorSettingHasBeenSet)
      CoCoA_ERROR(ERR::BadGlobalSettings, "allocator setting");
    if (arg1.myAllocatorSettingHasBeenSet) ans.mySetAllocatorSetting(arg1.myAllocatorSetting);
    if (arg2.myAllocatorSettingHasBeenSet) ans.mySetAllocatorSetting(arg2.myAllocatorSetting);

    if (arg1.mySliceSizeHasBeenSet && arg2.mySliceSizeHasBeenSet)
      CoCoA_ERROR(ERR::BadGlobalSettings, "GMPAllocator slice size");
    if (arg1.mySliceSizeHasBeenSet) ans.mySetSliceSize(arg1.mySliceSize);
    if (arg2.mySliceSizeHasBeenSet) ans.mySetSliceSize(arg2.mySliceSize);

    return ans;
  }


  const GlobalSettings UseSymmResidues(GlobalSettings().mySetResidueSetting(GlobalSettings::SymmResidues));
  const GlobalSettings UseNonNegResidues(GlobalSettings().mySetResidueSetting(GlobalSettings::NonNegResidues));
  const GlobalSettings UseSystemAllocatorForGMP(GlobalSettings().mySetAllocatorSetting(GlobalSettings::SystemAllocator));
  const GlobalSettings UseGMPAllocator(GlobalSettings().mySetAllocatorSetting(GlobalSettings::GMPAllocator));


  // ----------------------------------------------------------------------

  GlobalManager::ZZQQMgr::ZZQQMgr():
      myRingZZ(MakeUniqueInstanceOfRingZZ()),
      myRingQQ(MakeUniqueInstanceOfRingQQ(myRingZZ))
  {}

  GlobalManager::ZZQQMgr::~ZZQQMgr()
  {
    if (RingZZStillInUse(myRingZZ) || RingQQStillInUse(myRingQQ))
      std::cerr << std::endl
                << "*************************" << std::endl
                << "*** IMMINENT DISASTER ***" << std::endl
                << "*************************" << std::endl
                << std::endl
                << ">>>  CoCoA::GlobalManager being destroyed while CoCoA objects still live!  <<<" << std::endl
                << std::endl;
  }


  // ----------------------------------------------------------------------

  GlobalManager::GlobalManager(const GlobalSettings& settings):
      myResidueSetting(settings.myResidueSetting),
      myGMPMemMgr(settings.myAllocatorSetting, settings.mySliceSize),
      myZZQQMgr()
  {
// !!!***NOT THREAD SAFE***!!!  Must make next 3 lines atomic.
    // Complain if a GlobalManager object has already been created
    if (ourGlobalDataPtr != 0)
      CoCoA_ERROR(ERR::GlobalManager2, "GlobalManager ctor");
    ourGlobalDataPtr = this;
  }


  GlobalManager::~GlobalManager()
  {
// !!!***NOT THREAD SAFE***!!!  Must make check on ourGlobalDataPtr exclusive!
    if (ourGlobalDataPtr == 0)
      CoCoA_ERROR(ERR::SERIOUS, "GlobalManager dtor");
    // Delete registered globals in reverse order
    for (long i=len(myDtorList)-1; i >= 0; --i)
      myDtorList[i].myDtor(myDtorList[i].myPtr);
    ourGlobalDataPtr = 0; // "deregister" the global data
  }


//   bool DefaultResiduesAreSymm()
//   {
//     return GlobalManager::ptr("DefaultResiduesAreSymm")->myResiduesAreSymm;
//   }

  GlobalSettings::ResidueSetting DefaultResidueSetting()
  {
    return GlobalManager::ptr("DefaultResidueSetting")->myResidueSetting;
  }

  
  GlobalManager::PtrDtor::PtrDtor(void* ptr, void (*dtor)(void*)): myPtr(ptr), myDtor(dtor) {}

  void RegisterDtorForGlobal(void* ptr, void (*dtor)(void*))
  {
    GlobalManager::ptr("RegisterDtorForGlobal")->myDtorList.push_back(GlobalManager::PtrDtor(ptr,dtor));
  }


  //----------------------------------------------------------------------
  // pre-computed power list for univariate Hilbert-Poincare Series
  void MakeGlobalHPPowerList(const DenseUPolyRing& P)
  {  
    MakeHPPowerList(GlobalManager::ptr("MakeGlobalHPPowerList")->myHPPowerList,
                    P,
                    GlobalManager::ourHPMaxPower);
  }


  int HPPowerListMaxDeg()
  {  
    return GlobalManager::ptr("HPPowerListMaxDeg")->ourHPMaxPower;
  }

  
  ConstRefRingElem HPPowerList(int exp)
  {
    if (exp>HPPowerListMaxDeg())
      CoCoA_ERROR(ERR::ArgTooBig, "HPPowerList");
    return GlobalManager::ptr("HPPowerList")->myHPPowerList[exp];
  }
  
    
  void CopyHPPower(RingElem& res, int exp)
  {
    if (exp<=HPPowerListMaxDeg())
      res = HPPowerList(exp);
    else
    {
      res = HPPowerList(HPPowerListMaxDeg());
      const DenseUPolyRing HSRing = owner(res);
      for (long i=HPPowerListMaxDeg(); i<exp; ++i)
        HSRing->myMulBy1MinusXExp(raw(res), 1);
    }    
  }
  
  RandomSource& GlobalRandomSource()
  {
    return GlobalManager::ptr("GlobalRandomSource")->myRandomSource;
  }

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/GlobalManager.C,v 1.18 2014/07/09 13:01:17 abbott Exp $
// $Log: GlobalManager.C,v $
// Revision 1.18  2014/07/09 13:01:17  abbott
// Summary: Removed AsDenseUPolyRing
// Author: JAA
//
// Revision 1.17  2014/07/01 12:40:53  bigatti
// -- added CopyHPPower and argument check in HPPowerList
//
// Revision 1.16  2014/04/30 16:07:40  abbott
// Summary: Replaced X.size() by len(X)
// Author: JAA
//
// Revision 1.15  2013/06/17 08:54:02  abbott
// Added RegisterDtorForGlobal.
//
// Revision 1.14  2012/10/15 12:35:24  abbott
// Added  std::  prefix.
//
// Revision 1.13  2012/02/08 13:47:16  bigatti
// -- changed Z,Q --> ZZ,QQ
//
// Revision 1.12  2011/05/19 13:54:48  abbott
// Replaced DefaultResiduesAreSymm by DefaultResidueSetting.
//
// Revision 1.11  2011/05/03 10:03:32  abbott
// Added GlobalRandomSource.
// Internally added GlobalManager::ptr to allow neater implementations.
//
// Revision 1.10  2010/11/17 15:52:33  abbott
// Removed out-of-date include of GMPAllocator.H.
//
// Revision 1.9  2010/11/11 17:45:08  abbott
// Moved GMPMemMgr so that it is a nested class inside GlobalManager.
//
// Revision 1.8  2010/10/29 12:06:41  bigatti
// -- added globals for Hilbert-Poincare' series
//
// Revision 1.7  2010/10/27 20:58:45  abbott
// Major reorganization of GlobalManager and GMPAllocator.
//
// Revision 1.6  2010/10/22 14:03:04  abbott
// Major change to GMPAllocator -- it is now set/activated by the GlobalManager.
// This is a Friday afternoon check-in... hope to check in cleaner code in the
// next few days.
//
// Revision 1.5  2010/09/30 14:28:23  abbott
// Replaced auto_ptrs to RingZ and RingQ by direct values; ctor changed accordingly.
//
// Dtor now checks that ref counts in RingZ and RingQ are correct; if not, a rude
// message is printed on cerr (and the program will probably crash after the
// GlobalManager has been destroyed).
//
// Revision 1.4  2009/12/03 17:40:36  abbott
// Added include directives for ZZ.H (as a consequence of removing
// the directive from ring.H).
//
// Revision 1.3  2009/05/14 09:39:29  abbott
// Added possibility to specify "symmetric" or "non-negative" residues
// in quotients of ZZ.  Affects printing of elements in quotients of ZZ
// (also changed printing of elements in general quotient rings).
// Consequent changes in several tests.
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.2  2007/03/05 21:33:13  cocoa
// Improved/cleaned GlobalManager; added doc too.
//
// Revision 1.1  2007/03/03 14:02:11  bigatti
// -- "foundations" renamed into "GlobalManager"
//
// Revision 1.1  2007/03/02 16:46:28  cocoa
// New foundations object which calls ctors and dtors of global objects.
//
