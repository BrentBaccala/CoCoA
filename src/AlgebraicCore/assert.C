//   Copyright (c)  2005  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/assert.H"
#include "CoCoA/error.H"

#include <iostream>
using std::cerr;
using std::endl;
#include <sstream>
using std::ostringstream;


namespace CoCoA
{

  void AssertionFailed(const char* const cond, const char* const file, unsigned long line)
  {
    ostringstream message;
    message << endl
            << "===========================================================================" << endl
            << "===========================================================================" << endl
            << "==== CoCoA Assertion failed: [[" << cond << "]]" << endl
            << "==== File: " << file << endl
            << "==== Line: " << line << endl
            << "===========================================================================" << endl
            << "===========================================================================" << endl
            << endl;
    cerr << message.str();

    // Throw a CoCoA error in an unusual way... (so that the "right" file and line no. info appear).
    ThrowError(ErrorInfo(ERR::AssertFail, "AssertionFailed", file, line));
  }

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/assert.C,v 1.3 2010/12/17 16:11:06 abbott Exp $
// $Log: assert.C,v $
// Revision 1.3  2010/12/17 16:11:06  abbott
// Output related to assertions is now on standard C++ streams (rather
// than GlobalErrput, etc).
//
// Revision 1.2  2007/10/30 17:14:07  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.1  2006/10/06 14:04:57  cocoa
// The new assert header and implementation files.
// A new test.
//
//
