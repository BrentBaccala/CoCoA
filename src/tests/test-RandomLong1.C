//   Copyright (c)  2010  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/random.H"
#include "CoCoA/assert.H"  // for using TEST_ASSERT
#include "CoCoA/error.H"

#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
#include <vector>
using std::vector;

using namespace CoCoA;

// Handy macro for making assertions.
#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)

void program()
{
  GlobalManager CoCoAFoundations;

  const int N = 10;
  RandomSeqLong RndLong(-N,N);
  cout << "RndLong=" << RndLong << endl;

  const int ExpectedFreq = 256; // should be larger than 100ish
  vector<int> histo(2*N+1);
  for (int i=0; i < ExpectedFreq*(2*N+1); ++i)
  {
    const long rnd = NextValue(RndLong);
    TEST_ASSERT(rnd >= -N && rnd <= N);
    ++histo[N+rnd];
  }

  // Check histogram: each entry is > 0, and greatest and least freqs are within a factor of 2
  int MostFreq = histo[0];
  int LeastFreq = histo[0];
  for (int i=0; i < 2*N+1; ++i)
  {
    TEST_ASSERT(histo[i] != 0);
    if (histo[i] > MostFreq) { MostFreq = histo[i]; continue; }
    if (histo[i] < LeastFreq) { LeastFreq = histo[i]; }
  }
  const double ratio = double(MostFreq)/LeastFreq;
  TEST_ASSERT(ratio < 2.0);

  cout << "RndLong=" << RndLong << endl;
}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
