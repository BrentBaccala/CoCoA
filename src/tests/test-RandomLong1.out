RndLong=RandomSeqLong(lwb=-10, upb=10, seed=0, counter=0)
RndLong=RandomSeqLong(lwb=-10, upb=10, seed=0, counter=5376)
